### ISSUE TITLE
Adding a source does not seem to add it to the store

### ISSUE THREAD

**Issue by fricklerhandwerk** [andir/npins#57]
When I add a large repo as a source, it has to download for a long time. Using that source in a derivation requires downloading it again, apparently. Is that a user error or is it how the tool works?

**Comment by piegamesde** [comment]
I think that we try to use the correct Nix prefetch call to avoid this, but it might depend on the exact situation. Can you please show a command which adds a source which reproduces this?

**Comment by eclairevoyant** [comment]
I've run into a similar issue myself with the following pin:

```json
    "google-fonts": {
      "type": "Git",
      "repository": {
        "type": "GitHub",
        "owner": "google",
        "repo": "fonts"
      },
      "branch": "main",
      "revision": "3992c83a1afd00abb72a6c237159f20e419433dc",
      "url": "https://github.com/google/fonts/archive/3992c83a1afd00abb72a6c237159f20e419433dc.tar.gz",
      "hash": "1zcqm82c7a08mkkfhm7py494qa41q8pb7x237hdqas01qw5kzcp8"
    },
```

Updating the pin seems to take a while on its own - presumably the tarball must be downloaded to get the hash.
First build after updating the pin also takes a while, as it tries to redownload the tarball.
The tarball itself is massive, so the time spent is noticeable.

**Comment by fricklerhandwerk** [comment]
Well of course: `npins` [relies on `nix-prefetch-url`](https://github.com/andir/npins/blob/1e1495a5beeeb328f26dc1c34077969ff42e1546/src/nix.rs#L9-L18) and that in turn doesn't cache anything. When `builtins.fetchTarball` and such are called on the pinned URL, the whole download has to start over.

**Comment by fricklerhandwerk** [comment]
Could it be that this is because `npins` relies on `nix-prefetch-url`, which doesn't use the fetcher cache?

**Comment by piegamesde** [comment]
Yes indeed we shell out to the `nix-prefetch-url` and `nix-prefetch-git` commands (<https://github.com/andir/npins/blob/master/src/nix.rs>). I am a bit confused that these do not cache the result in the store. Are there better alternatives?

**Comment by fricklerhandwerk** [comment]
`nix-prefetch-url` produces a store object but doesn't use the fetcher cache, which means that `builtins.fetchurl` will always get a cache miss and only find that the resulting store object already exists after the source was downloaded.

**Comment by piegamesde** [comment]
Sounds like a nix bug to me tbh

**Comment by fricklerhandwerk** [comment]
I'd say it's a lack of deliberate design. A bug would mean something was intended but went wrong by accident. The code doesn't look like that.

**Comment by Lunaphied** [comment]
I would say this is an explicit bug, as the documentation for `nix-prefetch-url` claims

> If you don't know the hash, you would have to download the file first, and then fetchurl would download it again when you build your Nix expression. Since fetchurl uses the same name for the downloaded file as nix-prefetch-url, the redundant download can be avoided.

**Comment by eclairevoyant** [comment]
Per some testing I did, it seems that `nix-prefetch-*` does in fact cache, but the store path format is different from builtin fetchers. I see this issue was already fixed before, for a different fetcher (`nix-prefetch-git` vs `fetchGit`): https://github.com/andir/npins/commit/f0449d090cd4cbf17849d29821a53ead9a88c42a

So a similar fix for the remaining fetchers should do the trick.
